src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs

```diff
@@ -668,6 +670,91 @@ public static bool TryFormatSingle(float value, ReadOnlySpan<char> format, Numbe
             return null;
         }
 
+        public static string FormatHalf(Half value, string? format, NumberFormatInfo info)
+        {
+            var sb = new ValueStringBuilder(stackalloc char[CharStackBufferSize]);
+            return FormatHalf(ref sb, value, format, info) ?? sb.ToString();
+        }
+
+        /// <summary>Formats the specified value according to the specified format and info.</summary>
+        /// <returns>
+        /// Non-null if an existing string can be returned, in which case the builder will be unmodified.
+        /// Null if no existing string was returned, in which case the formatted output is in the builder.
+        /// </returns>
+        private static unsafe string? FormatHalf(ref ValueStringBuilder sb, Half value, ReadOnlySpan<char> format, NumberFormatInfo info)
+        {
+            if (!Half.IsFinite(value))
+            {
+                if (Half.IsNaN(value))
+                {
+                    return info.NaNSymbol;
+                }
+
+                return Half.IsNegative(value) ? info.NegativeInfinitySymbol : info.PositiveInfinitySymbol;
+            }
+
+            char fmt = ParseFormatSpecifier(format, out int precision);
+            byte* pDigits = stackalloc byte[HalfNumberBufferLength];
+
+            if (fmt == '\0')
+            {
+                precision = HalfPrecisionCustomFormat;
+            }
+
+            NumberBuffer number = new NumberBuffer(NumberBufferKind.FloatingPoint, pDigits, HalfNumberBufferLength);
+            number.IsNegative = Half.IsNegative(value);
+
+            // We need to track the original precision requested since some formats
+            // accept values like 0 and others may require additional fixups.
+            int nMaxDigits = GetFloatingPointMaxDigitsAndPrecision(fmt, ref precision, info, out bool isSignificantDigits);
+
+            if (((double)value != 0.0) && (!isSignificantDigits || !Grisu3.TryRunHalf(value, precision, ref number)))
```

Comment:
The following will do the right thing without involving a more costly conversion first:
```suggestion
            if ((value != default) && (!isSignificantDigits || !Grisu3.TryRunHalf(value, precision, ref number)))
```